### ISSUE TITLE
Flixel rendering overhaul

### ISSUE THREAD

**Issue by ninjamuffin99** [HaxeFlixel/flixel#2915]
https://haxeflixel.com/blog/08-openfl-5-compatibility-status-update/

Around 2017, Beeblrox posted some progress he made in redoing the way HaxeFlixel does things like shaders, and FlxStrip rework. By the title of that blog, that was on OpenFL 5. We are now on OpenFL 9! 

https://github.com/AustinEast/flixel/tree/Beeblerox-FlxMaterial-and-camera-scaling

Sometime around a year ago, I was getting Austin East to work on getting beeble's work running in modern flixel/haxe/openfl/lime

According to messages between me and austin:

> 2022/10/20
> ninjamuffin99: so where was the things with all the polygons / tris or whatev, thats with all the openGL rendering right?
> that includes like, making FlxStrip play nicer and all that?
> 
> austin: yeah all that high level API stuff is in :) like the stuff from the famous beeble blog post on the flixel website lol

and another conversation with austin about the project

> 
> okay so ive gone over everything and i want to break down all the big items and where theyre at. this will also be helpful for me since there is a lot going on here lol
> i originally started this as a port of beeble's branch, but things sorta evolved from there, based on conversations i had w him and looking at other renderers out there
> austineast — 05/29/2022 10:35 PM
> beeble's branch featured that fancy opengl renderer, but it was heavy tied to openfl, which has changed quite a bit since the branch was worked on and everything is basically broken. lol. at first i tried to fix it as-is, but after talking to beeble about it, he recommended rewriting it in a way that would be a bit closer to the metal (ie more opengl code through lime vs openfl) and pointed me to haxepunk's renderer, which ended up being a good resource
> austineast — 05/29/2022 10:42 PM
> i noticed haxepunk had a nice backend abstraction setup, which really helped keep the haxepunk codebase/renderer clean and easy to parse. its written in a way where any lime/openfl/nme specific code is separated out to a backends/ directory. and since i was in the process of planning out the flixel renderer rewrite, it seemed like a good idea to start doing that with flixel's codebase to help me really learn the internals and really future proof this new renderer
> austineast — 05/29/2022 11:06 PM
> with all that being said, heres the general statuses of the different pieces:
> 
> 1. merge beeble's branch into the latest flixel and get it compiling : complete (once it was compiling, i discovered that while the renderer abstraction was theoretically working well, openfl changes had really broken all of them hard. that lead me to the next task)
> 
> 2. port flixel's current working renderers (tiled mode/quad mode) to get them running with the new renderer abstraction: complete. ive tested it with a couple of my own projects and it seems to run fine with little-to-no breaking changes (it depended on whether the project was one of the ones where i do crazy graphics things or not lmao)
> 
> 3. separate openfl/lime code out into a backends directory: most (if not all) lime code has been moved out. there is still a bunch of openfl code. ive been porting things as i touch them, but there are just a lot of references to things like openfl's Bitmap and BitmapData  lol. luckily this backend stuff doesnt have to be done all at once, since its basically abstracting directly over openfl stuff like typedef Sprite = flash.display.Sprite; so it doesnt cause any compile errors. its not vital to get done immediately, but it will probably want to be finished before this gets merged into flixel 5.0+ though
> 
> 4. write a new opengl renderer: in progress. i have the basic architecture figured out (ie how to structure everything like vertex butffers, drawcalls for quads and tris, etc), but now im actually implementing the low level opengl part (ie managing the buffers, creating shaders and sending data to buffers correctly, sending it all to the gpu correctly etc) 
> a lot of time was spent on planning and figuring out the backend abstraction, but that was all on my own accord without really consulting you so im not factoring that into charged hours
> itll be really cool to have tho
> this opens the door for a heaps/kha/whatever backend in the future

### THE TLDR
- HaxeFlixel has less than stellar support for shaders (no multipass, cumbersome to toss in multiple textures, no camera buffer)
- In 2017, Beeblrox worked on fixing these things up, however OpenFL was moving a lot and making lots of changes
- In 2022, AustinEast worked off Beeblrox's old code and brought it up to modern standards of sorts, but needs work to bring it across the finish line.

### Where to go from here?
My gang at The Funkin' Crew can and will pay someone to put this work forward. Ideally.... it can be austin!! However he may be unavailable, so I suppose first course of action can be finding SOMEONE for me to throw money to.

**Comment by ACrazyTown** [comment]
I wonder if it's worth it to build an OpenGL renderer from scratch nowadays? It has been deprecated on Apple devices since 2018 and [Android will soon support it only through a compatibility layer](https://developer.android.com/games/develop/vulkan/overview#android-angle-on-vulkan-roadmap). 
Would having a custom renderer also make it harder to port to consoles? Could Flixel just use OpenFL's Stage3D/Context3D apis and leave the low level rendering parts to OpenFL/Lime?

**Comment by Fancy2209** [comment]
Thinking about this, maybe it'd make sense to use Starling instead of a from scratcch GL Renderer or Stage3D directly?
Starling would already allow better shader support trough Effects and wouldn't be as hard to port the codebase to from what I've researched.

**Comment by ACrazyTown** [comment]
I'll write this here as I think it's the most relevant issue, albeit not necessarily related to the Beeblerox branch. One step towards a proper renderer overhaul Flixel could take now would be to abstract away all the actual rendering bits from `FlxCamera`. This would make adding in a new backend a lot less painful than it would be now.

The Beeblerox branch accomplished this by giving each camera a unique `FlxCameraView` instance, which essentially just implements all the rendering related methods. All rendering related functions from `FlxCamera` (such as `drawPixels`) were moved into said view, and the camera just calls them. With something like this set in place, adding a new renderer is easy, just create a new camera view class and implement the methods.

There's also some methods in `CameraFrontEnd` that should be moved into `FlxCamera`, for example `lock()` and `unlock()`: 
https://github.com/HaxeFlixel/flixel/blob/f642a23c9b6106827fb18ae9bb9c13f6682d914e/flixel/system/frontEnds/CameraFrontEnd.hx#L256-L295
Instead of handling it here, it should really just be moved into a `camera.unlock()` / `camera.lock()` method and then have `CameraFrontEnd` call that.

**Comment by ACrazyTown** [comment]
I'll continue stealing this thread to spill some more of my thoughts on steps needed for a general renderer overhaul. After a bit of thinking, I think it could be possible to accomplish all of this without any breaking changes!

## 1. The Renderer Overhaul
I've talked about this a bit more in my comment above. This is the first and most important step!

To summarize:
- Move away all of the rendering bits from `FlxCamera` into a new base class that is then extended and implemented by each renderer. This could just be a copy of Beeble's [`FlxCameraView`](https://github.com/AustinEast/flixel/blob/Beeblerox-FlxMaterial-and-camera-scaling/flixel/system/render/common/FlxCameraView.hx) implementation!
- Create a foundation for other renderers to build off of, such as repurposing the drawQuads exclusive FlxDrawItem into a general FlxDrawCommand class.

This is something that could be done relatively easily, and could be squeezed into a minor version as there wouldn't be any API changes (any moved methods/properties can be kept as a shortcut to where they were moved).

## 2. FlxMaterial, FlxTrianglesData and alike
Beeble's branch added a couple of new classes that allowed for more fine grain control and optimization over things. It's a little tricky as these classes also require changes to the function signature of pretty much all rendering functions, but I think it's possible to avoid breaking changes with **overloads**. We can keep the same function but with different signatures, deprecate the old one and eventually get rid of it completely. It might not be the "cleanest" solution, but it works, and it'd allow us to also add this in a minor version!

## 3. New Renderer
This is where all the cool stuff comes into play! The current drawQuads renderer is less than ideal and lacks many standard rendering features (render textures, camera framebuffers and just generally more fine grain control over things). With the foundation set in place, this would mainly be a matter of just hooking everything up.

### OpenGL
Austin/Beeble's branch implemented a new renderer that went straight to the core and used **OpenGL** directly. This is as low level as we can get with the current tech, but I do have some doubts about it.

Using OpenGL directly would limit Flixel's ability to benefit from graphics backend changes in OpenFL. This means that if OpenFL were to ever adopt a new graphics backend such as Vulkan, or any proprietary graphics API used by consoles, Flixel would also need a new renderer to take advantage of the new APIs. Since console APIs are private, this would mean there'd have to be private Flixel forks that implement these APIs and need to be maintained seperately.

### Context3D
Another option to consider would be to use OpenFL's **Context3D** API. Flash introduced Context3D as a way to have hardware accelerated graphics, but in OpenFL it's essentially just a wrapper over OpenGL (on most platforms) with a slightly simpler API.

It's used under the hood by OpenFL, so it should be safe to assume that any new graphics backends (and consoles) would also work without any changes.

The issue with it is that the API was designed for Flash ~15 years ago. OpenFL made it work with OpenGL but there were no major public API changes, so it might pose certain limitations that'd require dropping down to the OpenGL level anyway.

## 4. All the other stuff
### Backends
Austin's branch did a lot more than just a rendering overhaul. There was work being done to remove the dependency on Lime/OpenFL and make it possible to hook up a different backend. This is really cool, but it can be done later and it isn't necessary for the rendering overhaul.

### Can we please kill renderBlit?
I've talked about [this](https://github.com/HaxeFlixel/flixel/issues/3468) before, but I really do think it should be done in the next major version. I think renderBlit is a hassle that just creates more issues than it's worth. Blitting is really only used by Flash (and HTML5 in some rare cases), so I don't see the value in maintaining a completely separate render method just for it, when we can take advantage of the existing drawQuads renderer instead.

---

I'll ping @Geokureli for his thoughts on this and how it should be contributed. I'm willing to start working on the first steps myself, and I already have a [little prototype of the renderer abstraction](https://github.com/ACrazyTown/flixel/tree/context3d) (though I'll probably have to redo it).

If anyone else has any insight or thoughts feel free to share! :D

**Comment by Geokureli** [comment]
## First and foremost: We lose more than flash by killing renderBlit.

Every now and then my html5 apps use it, for known and unknown reasons, and last I checked there's not much we can do to prevent a non wegbl context, especially depending on the users hardware. If we did kill it, we would need to throw a compile error for `<window hardware="false"/>` in project.xml and maybe have a default "your device/browser cannot play this game" state. This is NOT me saying "No, unconditionally. We must keep blit targets", I'm just saying that there's more at stake than just flash.

If we can allow different backends can we have a specific backend for blit targets? If our goal is to remove all the blit checks from code wouldn't that happen from changing all the `if (blit) camera.doBlit(this); else camera.doTiles(this);` calls. Most other non-trivial blit checks are to create some image buffer, or cache, to lower per-frame draw calls. I do think if we want to support different backends we should allow certain sprites the option to determine whether some sort of buffer is beneficial, and some way of doing that while staying somewhat agnostic to the specifics of various backends. There's probably some things in renderTile that also benefit from caching or buffering, too.

Lastly, Flash holds a personal spot in my heart, and the idea of dropping it gets me a little teary eyed. I also think it's possible for flash to make a comeback, via ruffle. But if it's literally half as much work to just gut out blit and start anew, and re-adding blit has a chance of killing this whole project, than let's gut it and come back to it later.

## Let's start abstracting lime/openfl now

Any places where a flixel dev uses a type defined in openfl/lime will need to be abstracted away, and we can benefit from that long before we completely overhaul everything. For instance, we abstracted openfl's confusing music streaming system from flixel's. We should do the same for BitmapDatas and Sounds. 

In particular, I want to abstract away openfl's confusing [image buffering](https://github.com/HaxeFlixel/flixel/pull/3345) and have a type that, by default, avoids unneeded ram and has a different tool or method for "image manipulation". I thought that was FlxMaterial's goal, but I still see public methods taking BitmapDatas, though I've never looked into Beeb's stuff extensively. I generally suck at reading lower-level rendering code, so I'm happy to see you take interest in this.

FlxCameraView is another great example, this seems like something we can work on iteratively, and see benefits in releases before the whole overhaul is complete. Given how many people went down this road and abandoned it, I think it's important to think of ways to do this without everything sitting in it's own branch for a long time. Merge early, merge often, break task down into smaller ones! If this happens over the course of multiple major version releases, that's fine.

## Lemme paper clip a pipe dream onto this

One of my long-term goal is to [make flixel more compositional](https://github.com/HaxeFlixel/flixel/issues/3053), specifically extracting all the rendering stuff from FlxSprite and putting it into some FlxImage class (maybe this is what FlxMaterial is doing?). this way any FlxObject or class extending FlxObject can be a "sprite"  by adding this component, much like how all the animation controllers where moved into `FlxAnimationController`, and how [collision will be extracted from FlxObect and put into a FlxCollider](https://github.com/HaxeFlixel/flixel/issues/3053?issue=HaxeFlixel%7Cflixel%7C2716). 

This is not mandatory, but if we're gonna break everything and reassemble the pieces, lets think about doing this, as well.

# Re: New Renderer

I assume context3d would be an easier transition between what we have now and going to pure OpenGL? How much harder would it be to skip this altogether? Both seem like a big enough rewrite that we may as well go straight to GL, but you know more, here.

> I think it's possible to avoid breaking changes with overloads. We can keep the same function but with different signatures, deprecate the old one and eventually get rid of it completely. It might not be the "cleanest" solution, but it works, and it'd allow us to also add this in a minor version!

Avoiding breaking changes with new methods and overloads is like my go-to, lol

Anyway, thanks for coming to my TED talk. btw, are you able to see this? They are marked private and I don't know what restrictions that causes

<img width="313" height="210" alt="Image" src="https://github.com/user-attachments/assets/bc3a098d-1e77-46e6-9a80-1885bb738e64" />

**Comment by ACrazyTown** [comment]
## Re: Re: renderBlit

I might've worded that poorly. I don't mean dropping Flash altogether, I mean replacing renderBlit with renderTile. OpenFL's Graphics API works with software renderers (including Flash), so with a few tweaks it can also work on Flash. However, considering that you've mentioned abstracting Lime/OpenFL, it may be a good idea to keep it so that it can work without the need of OpenFL (and its Graphics API).

## Re: Abstracting away Lime/OpenFL

> In particular, I want to abstract away openfl's confusing image buffering and have a type that, by default, avoids unneeded ram and has a different tool or method for "image manipulation". I thought that was FlxMaterial's goal, but I still see public methods taking BitmapDatas, though I've never looked into Beeb's stuff extensively.

This is something that has been on my mind as well. I think ideally we'd have 2 seperate image classes, a bitmap data that is editable and a texture that is not, and then `FlxGraphic` or whatever's used for the rendering would keep handles to both of those, depending on the use.

FlxMaterial is unrelated. My understanding is that it keeps data related to effects like shaders, blends and antialiasing.

## Re: Composition

No complaints here, I think that's a good idea and I'd like to see it happen!

## Re: Re: New Renderer

> I assume context3d would be an easier transition between what we have now and going to pure OpenGL?

Context3D and OpenGL are pretty much the same in terms of complexity, with Context3D being a bit simpler. The main benefit of Context3D would be the ability to benefit from graphics updates to OpenFL and their console backends. Since we've already mentioned abstracting away Lime/OpenFL, it might be better to bite the bullet and go straight for OpenGL then, since Context3D is Flash/OpenFL exclusive. 

> How much harder would it be to skip this altogether?

It wouldn't be hard but a new renderer is something Flixel could really benefit from. Better performance, better debug draw, better triangles/FlxStrip, more advanced rendering features like render textures, framebuffers, multitexture batching... is all something we could get from switching to GL directly instead of the current renderTile api.

---

> btw, are you able to see this? They are marked private and I don't know what restrictions that causes

No. I assume it means only members of the HaxeFlixel org can see them.

Well, with all this in mind, I think it might be worth trying to get Austin's branch up to date as it does a lot of the work already. Let's see how hard it is to merge 3+ years worth of upstream changes into a branch that changes a lot by itself

**Comment by ACrazyTown** [comment]
I'm going through the original OpenGL renderer branch, and there's a slight thing about the implementation that worries me. Some of the common classes directly reference the OpenGL renderer implementation. Ideally everything should be done through an abstraction.

Let's take [`FlxTrianglesData`](https://github.com/AustinEast/flixel/blob/Beeblerox-FlxMaterial-and-camera-scaling/flixel/graphics/FlxTrianglesData.hx) and [`FlxMaterial`](https://github.com/AustinEast/flixel/blob/Beeblerox-FlxMaterial-and-camera-scaling/flixel/graphics/FlxMaterial.hx) for example. Both classes have common, implementation agnostic code, but they also reference GL directly in some parts. IMO this is a bit unclean, and defeats the point of a renderer abstraction. I've been thinking about it a bit but I can't seem to figure out how to abstract this away without duplicating the common code.

It's not an urgent thing, but I think it'd be good to figure this out to future proof everything. If Flixel were to adopt any new renderer implementations things could start getting a bit messy. I think for now I could port the branch as-is (minus the breaking changes, so that we could fit this in a minor release) and we could leave cleaning everything up and polishing for a major version (so that we can also do breaking changes, if necessary). I'd appreciate any second thoughts here.
